### DISCUSSION TITLE
help needed writting sqlite for SHOM maps

### DISCUSSION THREAD

**Discussion by benoit35690** [osmandapp/OsmAnd#16494] [Maps Related]
Hello,

I know how to use the sqlite.db files, i have already completed with the Navionic and the French IGN maps. But in that cases the sqlite.db file was provided.
https://tech.gamuza.fr/OSM-AND-ajouter-carte-IGN.html
https://anygis.ru/Web/Html/Osmand_en

Now, I would like to have the French marine maps SHOM
https://services.data.shom.fr/support/fr/geoservices/WMS

Could you please give me indications how to write the sqlite.db file that matches ?
I have not understood how it works and which parameters are required to communicate with the WMS server.

Thanks

Benoit

**Comment by mariush444** [comment]
in theory it is quite easy
the method is explained 
https://docs.geoserver.org/2.22.x/en/user/services/wms/reference.html
or
https://enterprise.arcgis.com/en/server/latest/publish-services/windows/communicating-with-a-wms-service-in-a-web-browser.htm

the first step is to know the capabilities (maps/layers and their parameters) of WMS server

so you should click on the first or second link of your 3rd link to know it
https://services.data.shom.fr/INSPIRE/wms/r?service=WMS&version=1.3.0&request=GetCapabilities

please noticed **request=GetCapabilities**

This is xml file. Then you see what you can get.

You should build request that contains server name, request  **GetMap** and parameters

in your case
server
`https://services.data.shom.fr/INSPIRE/wms/r?`
request 
`service=WMS&version=1.3.0&request=GetMap`
and parameters. The list of parameters less or more are usually the same but you should know the value of this parameters.
You should read them from xml what map layers  do you need, what is the styles for it and other.

So for the example I choose one of the last map then I can build the request
`https://services.data.shom.fr/INSPIRE/wms/r?service=WMS&version=1.3.0&request=GetMap&CRS=EPSG:4326&Layers=TYPE_MAREE_PYR_PNG_3857_WMSR&bbox={bbox}&format=image/png&Styles=types_maree&width=256&height=256`

but osmand has some problems https://github.com/osmandapp/OsmAnd/issues/4201 with ver.1.3.0 
You can see that osmand downloaded something but most probably the georeferencing (lat,lon,bbox,crs) is wrong


in such cases I use **_**oruxmap**_** that is much much better for WMS services than osmand and helps to set parameters to use WMS automatically without guessing.

**Reply by benoit35690** [reply]
Thank you very much for all these information, I will take time to study it carrefully

**Comment by quantenschaum** [comment]
AFAIK OsmAnd can only display vector data when packaged as [OBF](https://osmand.net/docs/technical/osmand-file-formats/osmand-obf/). Online maps can be included if they are raster based and provided as [slippy tiles](https://en.wikipedia.org/wiki/Tiled_web_map) using the [Online Maps Plugin](https://osmand.net/docs/user/map/raster-maps). So you can access TMS and WMTS but not WMS directly. To access a WMS, you need some kind of proxy that translates WMS to TMS, like https://mapproxy.org/. You can add online maps to OsmAnd very easily via the [magic URL](https://osmand.net/docs/user/map/raster-maps#magic-url-to-install-map-source).

example: just click this to add [`https://anygis.ru/api/v1/Navionics_Marine_Charts_layer/{1}/{2}/{0}`](http://osmand.net/add-tile-source?name=Navionics&min_zoom=2&max_zoom=18&url_template=https://anygis.ru/api/v1/Navionics_Marine_Charts_layer/{1}/{2}/{0}) to OsmAnd (probably not legal :grin:)

The charts by SHOM are provided via a WMS, so they cannot be used in OsmAnd directly.

**Comment by quantenschaum** [comment]
Actually SHOM nautical charts are available freely via WMTS at `https://services.data.shom.fr/clevisu/wmts?layer=RASTER_MARINE_3857_WMTS&style=normal&tilematrixset=3857&Service=WMTS&Request=GetTile&Version=1.0.0&Format=image%2Fpng&TileMatrix={z}&TileCol={x}&TileRow={y}`. The caveat is, that you have to set a `Referer: https://data.shom.fr/` header in the request to make it work. OsmAnd does not support that.

**Reply by quantenschaum** [reply]
Well, actually can set the referer but not in the app. You can prepare a `sqlitedb` and set the `referer` field in the `info` table and osmand will send this referer.

```sql
BEGIN TRANSACTION;
CREATE TABLE IF NOT EXISTS "info" (
	"tilenumbering"	,
	"minzoom"	,
	"maxzoom"	,
	"title"	TEXT,
	"url"	TEXT,
	"randoms"	TEXT,
	"ellipsoid"	TEXT,
	"inverted_y"	TEXT,
	"referer"	TEXT,
	"useragent"	TEXT,
	"timecolumn"	TEXT,
	"expireminutes"	TEXT
);
CREATE TABLE IF NOT EXISTS "tiles" (
	"x"	int,
	"y"	int,
	"z"	int,
	"s"	int,
	"image"	blob,
	PRIMARY KEY("x","y","z","s")
);
INSERT INTO "info" VALUES ('simple',5,20,'SHOM','https://services.data.shom.fr/clevisu/wmts?layer=RASTER_MARINE_3857_WMTS&style=normal&tilematrixset=3857&Service=WMTS&Request=GetTile&Version=1.0.0&Format=image/png&TileMatrix={0}&TileCol={1}&TileRow={2}',NULL,'0','0','https://data.shom.fr/','Mozilla/5.0 AppleWebKit/537.36 Chrome/120.0.0.0 Safari/537.36','no','-1');
CREATE UNIQUE INDEX IF NOT EXISTS "IND" ON "tiles" (
	"x",
	"y",
	"z",
	"s"
);
COMMIT;
```

**Comment by quantenschaum** [comment]
https://mapproxy.org/ does allow setting of custom headers. 

Another interesting service for WMS-TMS-translation is http://whoots.mapwarper.net/

try this: `http://whoots.mapwarper.net/tms/{0}/{1}/{2}/6,5,4,3,2,1,0/https://gis.charttools.noaa.gov/arcgis/rest/services/MCS/ENCOnline/MapServer/exts/MaritimeChartService/WMSServer`
